### PR TITLE
GC single threadpool

### PR DESCRIPTION
**Pull request by ashtul**
This PR replaces 1 thread per index mechanism with a single threadpool.

### CODE REVIEW

**Comment by ashtul** [comment]
fix #524

File: src/gc.c [hunk 1]
```diff
@@ -62,44 +68,104 @@ GCContext* GCContext_CreateGC(RedisModuleString* keyName, float initialHZ, uint6
...
-    // Its better to just avoid it.
-    return 1;
+static void stopGC(GCContext* gc) {
+  gc->isStopped = 1;
+  if (gc->callbacks.kill) {
+    gc->callbacks.kill(gc->gcCtx);
+  }  
+}
+
+void GCContext_Timer_PeriodicCallback(RedisModuleCtx* ctx, void* data);
```
Comment (mnunberg) [inline src/gc.c]: static void

File: src/gc.c [hunk 2]
```diff
@@ -62,44 +68,104 @@ GCContext* GCContext_CreateGC(RedisModuleString* keyName, float initialHZ, uint6
...
+  gc->callbacks.onTerm(gc->gcCtx);
+  gc->gcCtx = NULL;
+  rm_free(gc);
+}
+
+static void internal_PeriodicCallback(void* data) {
+  GCContext* gc = data;
+  RedisModuleCtx* ctx = RSDummyContext;
+
+  if (gc->isStopped == 1) {
```
Comment (mnunberg) [inline src/gc.c]: why not `if (gc->isStopped)`? no need for strict numeric comparison

File: src/gc.c [hunk 3]
```diff
@@ -62,44 +68,104 @@ GCContext* GCContext_CreateGC(RedisModuleString* keyName, float initialHZ, uint6
...
+    // Its better to just avoid it.
+    GCContext* gc = data;
+    gc->timerID = scheduleNext(ctx, gc);
+    return;
+  }
+  thpool_add_work(gcThreadPools_g, internal_PeriodicCallback, data);
+}
+
 void GCContext_Start(GCContext* gc) {
   struct timespec interval = gc->callbacks.getInterval(gc->gcCtx);
```
Comment (mnunberg) [inline src/gc.c]: This is unused

File: src/gc.c [hunk 4]
```diff
@@ -113,10 +179,34 @@ void GCContext_OnDelete(GCContext* gc) {
...
+  GCContext_ForceBGInvoke(gc);
 }
 
 void GCContext_ForceBGInvoke(GCContext* gc) {
-  RMUtilTimer_Signal(gc->timer);
+  if (gc->isStopped == 1) return;
+
+  void* gcCtx;
+  RedisModuleCtx* ctx = RSDummyContext;
+  if (RedisModule_StopTimer(ctx, gc->timerID, &gcCtx) == REDISMODULE_OK) {
```
Comment (mnunberg) [inline src/gc.c]: Is there any condition in which the return value is not REDISMODULE_OK? also, why the assert?

File: src/gc.h [hunk 5]
```diff
@@ -4,12 +4,15 @@
...
 #include "redismodule.h"
 #include "util/dllist.h"
+#include "dep/thpool/thpool.h"
 #include <time.h>
 
 #ifdef __cplusplus
 extern "C" {
 #endif
 
+extern threadpool gcThreadPools_g;
```
Comment (mnunberg) [inline src/gc.h]: why extern and not static? is this accessed outside gc.c?

File: src/legacy_gc.c [hunk 6]
```diff
@@ -431,13 +431,13 @@ int GC_PeriodicCallback(RedisModuleCtx *ctx, void *privdata) {
 void GC_OnTerm(void *privdata) {
   GarbageCollectorCtx *gc = privdata;
   RedisModuleCtx *ctx = RedisModule_GetThreadSafeContext(NULL);
-  RedisModule_ThreadSafeContextLock(ctx);
+  //RedisModule_ThreadSafeContextLock(ctx);
```
Comment (mnunberg) [inline src/legacy_gc.c]: remove line

File: docs/Configuring.md [hunk 7]
```diff
@@ -318,3 +318,23 @@ $ redis-server --loadmodule ./redisearch.so GC_POLICY FORK FORK_GC_CLEAN_THRESHO
 
 * only to be combined with `GC_POLICY FORK`
 * added in v1.4.16
+
+
+## FORK_GC_THREAD_POOL_SIZE
```
Comment (mnunberg) [inline docs/Configuring.md]: 'threadpool' as one word feels more natural to me, but not a biggie

File: docs/Configuring.md [hunk 8]
```diff
@@ -318,3 +318,23 @@ $ redis-server --loadmodule ./redisearch.so GC_POLICY FORK FORK_GC_CLEAN_THRESHO
...
+
+### Example
+
+```
+$ redis-server --loadmodule ./redisearch.so GC_POLICY FORK FORK_GC_THREAD_POOL_SIZE 4
+```
+
+### Notes
+
+* only to be combined with `GC_POLICY FORK`
```
Comment (mnunberg) [inline docs/Configuring.md]: `GC_POLICY FORK` is the default

File: src/cpptests/t_forkgc.cpp [hunk 9]
```diff
@@ -14,6 +14,32 @@ static timespec getTimespecCb(void *) {
...
 }
 
+typedef struct {
+  RedisModuleCtx *ctx;
+  void *fgc;
+  IndexSpec *sp;
+} args_t;
+
+static bool runLoop;
+pthread_t thread;
```
Comment (mnunberg) [inline src/cpptests/t_forkgc.cpp]: static!

File: src/cpptests/t_forkgc.cpp [hunk 10]
```diff
@@ -28,11 +54,12 @@ class FGCTest : public ::testing::Test {
 
     // Set the interval timer to something lower, so we don't wait too
     // long
+    /*
```
Comment (mnunberg) [inline src/cpptests/t_forkgc.cpp]: remove, don't comment

File: src/cpptests/t_forkgc.cpp (hunk 9 above)
Comment (mnunberg) [reply src/cpptests/t_forkgc.cpp]: you actually don't need this, pthread_t just contains the thread id. if you're detaching it, you can just use it locally in the function.

File: src/cpptests/t_forkgc.cpp [hunk 11]
```diff
@@ -14,6 +14,32 @@ static timespec getTimespecCb(void *) {
...
+  void *fgc;
+  IndexSpec *sp;
+} args_t;
+
+static bool runLoop;
+pthread_t thread;
+
+void *cbWrapper(void *args) {
+  args_t *fgcArgs = (args_t *)args;
+  sleep(1);
```
Comment (mnunberg) [inline src/cpptests/t_forkgc.cpp]: no sleep, if you need some kind of timing, synchronize that in some other way.

File: src/cpptests/t_forkgc.cpp (hunk 9 above)
Comment (mnunberg) [reply src/cpptests/t_forkgc.cpp]: But it's probably a good idea to only allow one such thread to run at a time anyway

File: src/cpptests/t_forkgc.cpp [hunk 12]
```diff
@@ -14,6 +14,32 @@ static timespec getTimespecCb(void *) {
   return ts;
 }
 
+typedef struct {
+  RedisModuleCtx *ctx;
+  void *fgc;
+  IndexSpec *sp;
+} args_t;
+
+static bool runLoop;
```
Comment (mnunberg) [inline src/cpptests/t_forkgc.cpp]: you don't seem to be using this variable?

File: src/cpptests/t_forkgc.cpp [hunk 13]
```diff
@@ -14,6 +14,32 @@ static timespec getTimespecCb(void *) {
   return ts;
 }
 
+typedef struct {
+  RedisModuleCtx *ctx;
```
Comment (mnunberg) [inline src/cpptests/t_forkgc.cpp]: RSDummyContext

File: src/gc.c [hunk 14]
```diff
@@ -62,44 +68,104 @@ GCContext* GCContext_CreateGC(RedisModuleString* keyName, float initialHZ, uint6
...
+  RedisModuleBlockedClient* bClient = BlockClients_pop(&gc->bClients);
+  if (bClient) {
+    RedisModule_UnblockClient(bClient, NULL);
+  }
+  gc->callbacks.onTerm(gc->gcCtx);
+  gc->gcCtx = NULL;
+  rm_free(gc);
+}
+
+static void internal_PeriodicCallback(void* data) {
```
Comment (mnunberg) [inline src/gc.c]: `threadCallback`; of course it's "internal" if it's static

File: src/gc.c [hunk 15]
```diff
@@ -62,44 +68,104 @@ GCContext* GCContext_CreateGC(RedisModuleString* keyName, float initialHZ, uint6
...
+    return;
   }
-  return ret;
+  gc->timerID = scheduleNext(ctx, gc);
+  RedisModule_ThreadSafeContextUnlock(ctx);
 }
 
-static void GCContext_OnTerm(void* privdata) {
-  GCContext* gc = privdata;
+static void internal_DestroyCallback(void* data) {
```
Comment (mnunberg) [inline src/gc.c]: `destroyCallback`, ditto

File: src/gc.c [hunk 16]
```diff
@@ -62,44 +68,104 @@ GCContext* GCContext_CreateGC(RedisModuleString* keyName, float initialHZ, uint6
...
+  RedisModuleBlockedClient* bClient;
+  while ((bClient = BlockClients_pop(&gc->bClients))) {  
+    RedisModule_UnblockClient(bClient, NULL);
+  }
+  
   gc->callbacks.onTerm(gc->gcCtx);
   rm_free(gc);
 }
 
+void GCContext_Timer_PeriodicCallback(RedisModuleCtx* ctx, void* data) {
```
Comment (mnunberg) [inline src/gc.c]: `timerCallback`; `static`

File: src/gc.h [hunk 17]
```diff
@@ -33,9 +36,11 @@ typedef struct GCCallbacks {
 
 typedef struct GCContext {
   void* gcCtx;
-  struct RMUtilTimer* timer;
+  // struct RMUtilTimer* timer;
+  RedisModuleTimerID timerID;
   BlockClients bClients;
   GCCallbacks callbacks;
+  int isStopped;
```
Comment (mnunberg) [inline src/gc.h]: i think simply calling it `stopped` would be less assaulting on the eyes

File: src/gc.c (hunk 4 above)
Comment (ashtul) [reply src/gc.c]: `RedisModule_StopTimer` returns ERR if no item was found in the queue.
All remove. Remains of previous tests.

File: src/gc.h (hunk 5 above)
Comment (ashtul) [reply src/gc.h]: We made it configurable with FORK_GC_THREAD_POOL_SIZE flag

File: src/cpptests/t_forkgc.cpp (hunk 13 above)
Comment (ashtul) [reply src/cpptests/t_forkgc.cpp]: This is using the mock API.
In line 68: `RMCK::Context ctx;`

File: src/gc.h (hunk 5 above)
Comment (mnunberg) [reply src/gc.h]: Yeah, but the symbol itself is not accessed outside gc.c; `static` doesn't mean "not relevant to anything else in the code", it simply means whether other compilation units (=files) can have direct access to that symbol.

File: src/gc.c [hunk 18]
```diff
@@ -62,44 +67,98 @@ GCContext* GCContext_CreateGC(RedisModuleString* keyName, float initialHZ, uint6
...
 
+  RedisModule_ThreadSafeContextLock(ctx);
   RedisModuleBlockedClient* bClient = BlockClients_pop(&gc->bClients);
   if (bClient) {
     RedisModule_UnblockClient(bClient, NULL);
   }
-  if (gc->timer) {
-    // Timer could've been deleted..
-    RMUtilTimer_SetInterval(gc->timer, gc->callbacks.getInterval(gc->gcCtx));
+  if (!ret || gc->stopped) {
```
Comment (MeirShpilraien) [inline src/gc.c]: You are not checking that you have more blocked clients

File: src/gc.c [hunk 19]
```diff
@@ -62,44 +67,98 @@ GCContext* GCContext_CreateGC(RedisModuleString* keyName, float initialHZ, uint6
...
 
-static void GCContext_OnTerm(void* privdata) {
-  GCContext* gc = privdata;
+static void destroyCallback(void* data) {
+  GCContext* gc = data;
+  RedisModuleCtx* ctx = RSDummyContext;
+  assert(gc->stopped == 1);
+
+  RedisModule_ThreadSafeContextLock(ctx);
+  RedisModule_StopTimer(ctx, gc->timerID, NULL);
```
Comment (MeirShpilraien) [inline src/gc.c]: What if you do not have schedule timer here?

File: src/gc.c [hunk 20]
```diff
@@ -113,10 +172,31 @@ void GCContext_OnDelete(GCContext* gc) {
...
+
   BlockClients_push(&gc->bClients, bc);
-  RMUtilTimer_ForceInvoke(gc->timer);
+  GCContext_ForceBGInvoke(gc);
 }
 
 void GCContext_ForceBGInvoke(GCContext* gc) {
-  RMUtilTimer_Signal(gc->timer);
+  RedisModuleCtx* ctx = RSDummyContext;
+  RedisModule_StopTimer(ctx, gc->timerID, NULL);
```
Comment (MeirShpilraien) [inline src/gc.c]: You might not have a timer here to stop

Comment (ashtul) [reply src/gc.c]: In that case, it returns RM_ERR but it is legal.

File: src/gc.c (hunk 19 above)
Comment (ashtul) [reply src/gc.c]: See below

File: src/gc.c (hunk 18 above)
Comment (ashtul) [reply src/gc.c]: since we can have a single blocked client at the time as we add tasks with DEBUG command

Comment (MeirShpilraien) [reply src/gc.c]: Two clients can run the debug command at the same time

File: src/gc.c [hunk 21]
```diff
@@ -109,7 +110,10 @@ static void threadCallback(void* data) {
     RedisModule_ThreadSafeContextUnlock(ctx);
     return;
   }
-  gc->timerID = scheduleNext(gc);
+  // schedule next if not debug 
+  if (!bClient) {
```
Comment (MeirShpilraien) [inline src/gc.c]: You need to schedule next if &gc->bClients is empty if you had a blocked client but now you do not have any more you still need to schedule, starting to think we really should put the blocked client as an argument to the task and remove gc->bClients

Comment (ashtul) [reply src/gc.c]: @MeirShpilraien 
bClient is NULL only if the list is empty

Comment (MeirShpilraien) [reply src/gc.c]: @ashtul it might be that this the last bc in this case you need to trigger the timmer but you will not and the GC will stop running...

**Comment by ashtul** [comment]
replaced b y #1143

**Comment by ashtul** [comment]
replace by #1154

**Pull request closed without merging**